execution-engine/src/wasi/wasmtime.rs

```diff
@@ -0,0 +1,720 @@
+//! An implementation of the ExecutionEngine runtime state for Wasmtime.
+//!
+//! ## Authors
+//!
+//! The Veracruz Development Team.
+//!
+//! ## Copyright
+//!
+//! See the file `LICENSE.markdown` in the Veracruz root directory for licensing
+//! and copyright information.
+
+use crate::{
+    fs::FileSystem,
+    wasi::common::{
+        EntrySignature, ExecutionEngine, FatalEngineError, HostFunctionIndexOrName, MemoryHandler,
+        WASIAPIName, WASIWrapper,
+    },
+};
+use lazy_static::lazy_static;
+#[cfg(any(feature = "std", feature = "tz", feature = "nitro"))]
+use std::sync::{Arc, Mutex};
+#[cfg(feature = "sgx")]
+use std::sync::{Arc, SgxMutex as Mutex};
+use std::{collections::HashMap, convert::TryFrom, vec::Vec};
+use veracruz_utils::policy::principal::Principal;
+use wasi_types::ErrNo;
+use wasmtime::{Caller, Extern, ExternType, Func, Instance, Module, Store, ValType};
+
+////////////////////////////////////////////////////////////////////////////////
+// The Wasmtime runtime state.
+////////////////////////////////////////////////////////////////////////////////
+
+lazy_static! {
+    // The initial value has NO use.
+    static ref VFS_INSTANCE: Mutex<WASIWrapper> = Mutex::new(WASIWrapper::new(Arc::new(Mutex::new(FileSystem::new(HashMap::new()))), Principal::NoCap));
+}
+
+/// A macro for lock the global VFS and store the result in the variable,
+/// which will be captured by `$vfs` in the macro.
+/// If the locks fails, it returns Busy error code.
+macro_rules! lock_vfs {
+    () => {
+        match VFS_INSTANCE.lock() {
+            Ok(v) => v,
+            Err(_) => return ErrNo::Busy as u32,
+        }
+    };
+}
+
+/// A macro to call try_from on a untyped raw wasi parameter `$var`.
+/// It converts `$var` to the typed version, `$t`, of the parameter as a value,
+/// or returns from the function with the `Inval` error code.
+macro_rules! convert_wasi_arg {
+    ($var:ident, $t:ty) => {
+        match <$t>::try_from($var) {
+            Err(_) => return ErrNo::Inval as u32,
+            Ok(o) => o,
+        }
+    };
+}
+
+/// Impl the MemoryHandler for Caller.
+/// This allows passing the Caller to WASIWrapper on any VFS call.
+impl<'a> MemoryHandler for Caller<'a> {
```

Comment:
Why the lifetime parameter?

Reply:
The Caller type requires a  lifetime <'a>, it actually not use in our code.